From ffaf3d328d75a10f6f4344921f543a388b6a2e31 Mon Sep 17 00:00:00 2001
From: v1gnesh <v1gnesh@users.noreply.github.com>
Date: Wed, 13 Dec 2023 08:55:29 -0500
Subject: [PATCH] Add support for zOS

---
 extractor_unix.go | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/extractor_unix.go b/extractor_unix.go
index f4f2746..1cbdfd2 100644
--- a/extractor_unix.go
+++ b/extractor_unix.go
@@ -29,7 +29,13 @@ func lchmod(name string, mode os.FileMode) error {
 }
 
 func lchtimes(name string, mode os.FileMode, atime, mtime time.Time) error {
-	at := unix.NsecToTimeval(atime.UnixNano())
+  if runtime.GOOS == "zos" {
+    if err := lchmod(name, mode); err != nil {
+      return err
+    }
+  }
+
+  at := unix.NsecToTimeval(atime.UnixNano())
 	mt := unix.NsecToTimeval(mtime.UnixNano())
 	tv := [2]unix.Timeval{at, mt}
 
-- 
2.42.1

